packages/client/src/helpers/runWithRetry.ts

```diff
@@ -0,0 +1,87 @@
+const sleep = (time: number) =>
+  new Promise<void>((resolve) => {
+    setTimeout(resolve, time);
+  });
+
+export class RetryError extends Error {
+  public name = 'RetryError';
+
+  private static errorMap = {
+    abort: 'Value changed, retry handler aborted',
+    finished: 'Reached maximum amount of retries',
+  };
+
+  constructor({ type }: { type: keyof typeof RetryError.errorMap }) {
+    super(RetryError.errorMap[type]);
+  }
+}
+
+/**
+ * Function which wraps asynchronous functions with retry mechanism which'll keep executing said
+ * function pre-defined number of times until it resolves, rejects or the retry mechanism runs out of available attempts.
+ *
+ * #### Available options:
+ *  - `didValueChange` - check function with initial argument value to check against new values, return true to abort next attempt
+ *
+ *  - `isRetryable` - error evaluation function which determines whether to retry the execution
+ *
+ *  - `delayBetweenRetries` - accepts either fixed value or function which provides retry attempt as the argument, expects number as return value
+ *
+ *  - `retryAttempts` - number of attempts to try out before rejecting the promise
+ */
+export const runWithRetry =
+  <T extends (...functionArguments: any[]) => Promise<any>>(
+    f: T,
+    {
+      retryAttempts = 3,
+      delayBetweenRetries,
+      isRetryable,
+      didValueChange,
+    }: {
+      retryAttempts?: number;
+      delayBetweenRetries?: number | ((attempt: number) => number);
+      isRetryable: (error: unknown) => boolean;
+      didValueChange?: (...functionArguments: Parameters<T>) => boolean;
```

Comment:
I think all of these inputs shall be marked as optional in their type definitions but they should have initial defaults:
- `delayBetweenRetries = retryInterval`
- `isRetryable = () => true`
- `didValueChange = () => false`

What do you think about this?

Reply:
I have applied most of your suggestions but `delayBetweenRetries` is the only one I'm not sure about - I feel like we should keep the default state of the retry mechanism as `FastAndSimple` - `retryInterval` function as default would make it something between `FastAndSimple` and `NeverGonnaGiveYouUp`.

Reply:
Also - I'm not going to provide default functions to not pollute the stack with unnecessary function executions, we'll execute only if they're available - otherwise we skip.

The only change is to `isRetryable` where we try to execute it but if it does not exist we fall back to `?? true`.

Reply:
ok, sounds good.